mesonbuild/build.py

```diff
@@ -758,8 +758,8 @@ def __init__(
         self.link(link_targets)
         self.link_whole(link_whole_targets)
 
-        if not any([self.sources, self.generated, self.objects, self.link_whole_targets, self.structured_sources,
-                    kwargs.pop('_allow_no_sources', False)]):
+        if not any([[src for src in self.sources if not is_header(src)], self.generated, self.objects,
+                    self.link_whole_targets, self.structured_sources, kwargs.pop('_allow_no_sources', False)]):
```

Comment:
if not `any([(src for src in ...), ...])` Otherwise we're creating a concrete list, even if the first element ends up being True.

Reply:
> any([(src for src in ...), ...])

that doesn't seem to work

Reply:
Ohhhhhh. Yes. you would need do something like:
```python
if not any(itertools.chain((src for src in ...), self.generated, ...):
   ...
```
That's a lot more work than just making the concrete list. So It's up to you. If you like the list leave it, if you like itertools.chain do that.

Reply:
I think what you meant was: `if not any([*(src for src in self.sources if not is_header(src)), self.generated, self.objects,` (you need to unpack the generator).

Reply:
> Ohhhhhh. Yes. you would need do something like:
> if not any(itertools.chain((src for src in ...), self.generated, ...):

FYI this doesn't work either, as it fails with `TypeError: 'NoneType' object is not iterable` if something passed to it is None

Reply:
Okay, so those are allowed to be `None` rather than an empty list, which seems like bad design, but just stick with the list comprehension you had. If I ever get back to cleaning that stuff up I can change it then :)